rpc/dial.go

```diff
@@ -94,90 +101,137 @@ func dial(
+				// This path is also called by an mdns direct connection and ignores that case.
+				// This will skip all Authenticate/AuthenticateTo calls for the signaler.
+				if !dOpts.usingMDNS && dOpts.authMaterial == "" && dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial != "" {
+					logger.Debug("using signaling's external auth as auth material")
+					dOpts.authMaterial = dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial
+					dOpts.creds = Credentials{}
+				}
 			}
-			fixupWebRTCOptions(dOpts, target, port)
-
-			// When connecting to an external signaler for WebRTC, we assume we can use the external auth's material.
-			// This path is also called by an mdns direct connection and ignores that case.
-			// This will skip all Authenticate/AuthenticateTo calls for the signaler.
-			if !dOpts.usingMDNS && dOpts.authMaterial == "" && dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial != "" {
-				logger.Debug("using signaling's external auth as auth material")
-				dOpts.authMaterial = dOpts.webrtcOpts.SignalingExternalAuthAuthMaterial
-				dOpts.creds = Credentials{}
+
+			if dOpts.debug {
+				logger.Debugw(
+					"trying WebRTC",
+					"signaling_server", dOpts.webrtcOpts.SignalingServerAddress,
+					"host", originalAddress,
+				)
 			}
-		}
 
-		if dOpts.debug {
-			logger.Debugw(
-				"trying WebRTC",
-				"signaling_server", dOpts.webrtcOpts.SignalingServerAddress,
-				"host", originalAddress,
-			)
-		}
+			conn, cached, err := dialFunc(
+				wrtcCtx,
+				"webrtc",
+				fmt.Sprintf("%s->%s", dOpts.webrtcOpts.SignalingServerAddress, originalAddress),
+				buildKeyExtra(dOpts),
+				func() (ClientConn, error) {
+					return dialWebRTC(
+						wrtcCtx,
+						dOpts.webrtcOpts.SignalingServerAddress,
+						originalAddress,
+						dOpts,
+						logger,
+					)
+				})
+			if err == nil {
+				if !cached {
+					logger.Debug("connected via WebRTC")
+				} else if dOpts.debug {
+					logger.Debug("connected via WebRTC (cached)")
+				}
+				connCh <- dialSuccess{conn, cached, 1}
+				return
+			}
+			if !errors.Is(err, ErrNoWebRTCSignaler) {
+				errCh <- err
+				return
+			}
+			if wrtcCtx.Err() != nil {
+				errCh <- err
+				return
+			}
+			errCh <- nil
+		}()
+	}
 
-		conn, cached, err := dialFunc(
-			ctx,
-			"webrtc",
-			fmt.Sprintf("%s->%s", dOpts.webrtcOpts.SignalingServerAddress, originalAddress),
-			buildKeyExtra(dOpts),
-			func() (ClientConn, error) {
-				return dialWebRTC(
-					ctx,
-					dOpts.webrtcOpts.SignalingServerAddress,
-					originalAddress,
-					dOpts,
-					logger,
-				)
-			})
-		if err == nil {
-			if !cached {
-				logger.Debug("connected via WebRTC")
-			} else if dOpts.debug {
-				logger.Debug("connected via WebRTC (cached)")
+	var (
+		dialFails  int
+		dialErrors []error
+	)
+	for dialFails < len(cancels) {
+		select {
+		case success := <-connCh:
+			for i, cancel := range cancels {
+				if i != success.key {
+					cancel()
```

Comment:
looking at this again, I don't think this handles the case of the slower connection succeeding - my understanding is that `cancel` does nothing in that case, and the connection gets leaked, right?